From 9e5c6f774aeeec96ea63ecac3603389486e36580 Mon Sep 17 00:00:00 2001
From: Tristan Cacqueray <tdecacqu@redhat.com>
Date: Wed, 10 May 2017 10:35:53 +0000
Subject: [PATCH] Support jenkins-job-builder>=2

Recent version of jenkins-job-builder renamed the Builder class
to JenkinsManager and removed the embedded parser.

Change-Id: Iba78c0cdf1592126ec8244ae5424f6fbb5436c32
---
 requirements.txt                     |  1 +
 zuul/launcher/ansiblelaunchserver.py | 25 +++++++++++++++++++++----
 2 files changed, 22 insertions(+), 4 deletions(-)

diff --git a/requirements.txt b/requirements.txt
index 963c899..fdd77af 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -7,6 +7,7 @@ paramiko>=1.8.0,<2.0.0
 GitPython>=0.3.3,<2.1.2
 ordereddict
 python-daemon>=2.0.4,<2.1.0
+jenkins-job-builder
 extras
 statsd>=1.0.0,<3.0
 voluptuous>=0.10.2
diff --git a/zuul/launcher/ansiblelaunchserver.py b/zuul/launcher/ansiblelaunchserver.py
index 9a65da9..a50cdd3 100644
--- a/zuul/launcher/ansiblelaunchserver.py
+++ b/zuul/launcher/ansiblelaunchserver.py
@@ -29,7 +29,17 @@ import Queue
 
 import gear
 import yaml
-import jenkins_jobs.builder
+
+try:
+    from jenkins_jobs.builder import Builder as jenkins_jobs_builder
+    jjb_2 = False
+except ImportError:
+    from jenkins_jobs.builder import JenkinsManager as jenkins_jobs_builder
+    from jenkins_jobs.parser import YamlParser
+    from jenkins_jobs.config import JJBConfig
+    from jenkins_jobs.registry import ModuleRegistry
+    jjb_2 = True
+
 import jenkins_jobs.formatter
 import zmq
 
@@ -302,8 +312,13 @@ class LaunchServer(object):
         self.log.debug("Loading jobs")
         builder = JJB()
         path = self.config.get('launcher', 'jenkins_jobs')
-        builder.load_files([path])
-        builder.parser.expandYaml()
+        if jjb_2:
+            registry = ModuleRegistry(JJBConfig())
+            builder.parser.load_files([path])
+            builder.parser.expandYaml(registry, None)
+        else:
+            builder.load_files([path])
+            builder.parser.expandYaml()
         unseen = set(self.jobs.keys())
         for job in builder.parser.jobs:
             builder.expandMacros(job)
@@ -1541,9 +1556,11 @@ class NodeWorker(object):
         return ret == 0
 
 
-class JJB(jenkins_jobs.builder.Builder):
+class JJB(jenkins_jobs_builder):
     def __init__(self):
         self.global_config = None
+        if jjb_2:
+            self.parser = YamlParser(JJBConfig())
         self._plugins_list = []
 
     def expandComponent(self, component_type, component, template_data):
-- 
2.9.3

